src/validator/validator.ts

```diff
@@ -104,10 +104,10 @@ export class Validator {
     this.isArray = true
     const res = validator(this)
     const arr = new VArray(res, path)
+    this.isArray = false
```

Comment:
I reconsidered and think it would be better to create a new instance of `Validator`.

```ts
  array = <T extends Schema>(path: string, validator: (v: Validator) => T): VArray<T> => {
    const newValidator = new Validator()
    newValidator.isArray = true
    const res = validator(newValidator)
    const arr = new VArray(res, path)
    return arr
  }
  object = <T extends Schema>(path: string, validator: (v: Validator) => T): VObject<T> => {
    const newValidator = new Validator()
    newValidator.isArray = this.isArray
    const res = validator(newValidator)
    const obj = new VObject(res, path)
    return obj
  }
```

In this case, a different object from `this` should be invoked in `validator`.

Reply:
yes I totally agree!

To be honest I thought of this but I wasn't sure there was a specific reason why the same instance of `Validator` was being used (I guess it is more efficient, although I can't imagine it making a noticeable impact on performances).

I'll make the change :slightly_smiling_face::+1:

Reply:
@yusukebe I've changed the implementation, I soft of followed your suggestion but changed things up a bit, I thinks (please let me know if I'm wrong) that `isArray` should not be accessible from the outside of the class (as the user can access that as well).
So I made the `isArray` (I've renamed it to `inArray` as that makes more sense to me, please let me know if you want me to revert that) a private field of the class which you pass in the constructor. We also don't want the user to do that so I renamed the class and removed its export, what we are exporting then is a `Validator` class which wraps the old class but it removes the `inArray` parameter so that the user doesn't see that at all (since it is an implementation details I think).

What do you think? :slightly_smiling_face:

Reply:
Actually... I don't think that the `Validator` class is accessible from user code is it? in that case I think that the extra class is not really necessary...

Reply:
extra class removed

Reply:
@dario-piotrowicz 

Renaming `isArray` to `inArray` and changing its accessibility of it are great! The `Validator` class must be accessed from `middleware/validator/middleware.ts`, so I think it's OK to keep it.

I want to go with this current code, OK?

Reply:
Yes, if you're ok with it I think it should be good to get merged :smiley:

Reply:
OK!

In addition, the most important reason to export `Validator` is there are use cases that the developer can write validation schema outside of the `validator` function.

```ts
const schema = (v: Validator) => ({
  title: v.body('title').isRequired(),
  body: v.body('body').isRequired(),
})

app.post('/posts', validator(schema), (c) => {
  const { title, body } = c.req.valid()
  //...
})
```

Reply:
ah I see, thanks for clarifying :slightly_smiling_face::+1: